tests/lib/thrift/ThriftTest/src/client.cpp

```diff
@@ -0,0 +1,509 @@
+    zassert_equal(it26->second, Insanity(), NULL);
+  }
+}
+
+static void exception_test(void) {
+  ThriftTestClient& testClient = *(context.client);
+  try {
+    testClient.testException("Xception");
+    zassert_unreachable("Xception: Unhandled exception should abort execution");
+  } catch (Xception& e) {
+  }
+
+  try {
+    testClient.testException("TException");
+    zassert_unreachable("TException: Unhandled exception should abort execution");
+  } catch (const TException&) {
+  }
+
+  try {
+    testClient.testException("success");
+  } catch (exception& ex) {
+    zassert_exception("%s", ex.what());
+  }
+}
+
+static void multi_exception_test(void) {
+  ThriftTestClient& testClient = *(context.client);
+  try {
+    Xtruct result;
+    testClient.testMultiException(result, "Xception", "test 1");
+    zassert_unreachable("Xception: Unhandled exception should abort execution");
+  } catch (Xception& e) {
+  }
+
+  try {
+    Xtruct result;
+    testClient.testMultiException(result, "Xception2", "test 2");
+    zassert_unreachable("Xception2: Unhandled exception should abort execution");
+
+  } catch (Xception2& e) {
+  }
+
+  try {
+    Xtruct result;
+  } catch (exception& ex) {
+    zassert_exception("%s", ex.what());
+  }
+}
+
+static void oneway_void_test(void) {
+  ThriftTestClient& testClient = *(context.client);
+  uint64_t startOneway = now();
+  testClient.testOneway(1);
+  uint64_t elapsed = now() - startOneway;
+  zassert_true(elapsed <= 200 * 1000 /* 0.2 seconds */, "took %.2f ms", (double)elapsed / 1000.0);
+}
+
+static shared_ptr<ThriftTestClient> setup_client(const string& my_addr, const int port) {
+  std::shared_ptr<TTransport> trans(new TSocket(my_addr, port));
+  std::shared_ptr<TTransport> transport(new TBufferedTransport(trans));
+  std::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));
+  transport->open();
+  return std::make_shared<ThriftTestClient>(protocol);
+}
+
+static void* server_func(void* arg) {
+  (void)arg;
+
+  context.server->serve();
+
+  return nullptr;
+}
+
+static void setup(void) {
+  int rv;
+  pthread_attr_t attr;
+  pthread_attr_t* attrp = &attr;
+
+  if (IS_ENABLED(CONFIG_ARCH_POSIX)) {
+    attrp = NULL;
+  } else {
+    rv = pthread_attr_init(attrp);
+    zassert_equal(0, rv, "pthread_attr_init failed: %d", rv);
+    rv = pthread_attr_setstack(attrp, ThriftTest_server_stack, HELLO_SERVER_STACK_SIZE);
+    zassert_equal(0, rv, "pthread_attr_setstack failed: %d", rv);
+  }
+
+  constexpr int port = 4242;
+
+  // set up server
+  context.server = setup_server("0.0.0.0", port);
+
+  // start the server
+  rv = pthread_create(&context.server_thread, attrp, server_func, nullptr);
+  zassert_equal(0, rv, "pthread_create failed: %d", rv);
+
+  // set up client
+  context.client = setup_client("192.0.2.1", port);
+}
+
+static void teardown(void) {
+  void* unused;
+
+  context.server->stop();
+
+  pthread_join(context.server_thread, &unused);
+}
+
+void test_main(void) {
```

Comment:
It would probably make sense to put this into `main.cpp`.

I'm not 100% sure if `zassert_*()` works when not run as part of the test suite, so it's probably safer to use `ztest_unit_test_setup_teardown(test_xxx, setup, teardown)` even though it is slightly more verbose.

In the future, when we look at switching to the newer ztest API, there is a `setup()` for the suite, which would be a good place to open a socket, and then `teardown()` for tearing down the suite, but then also `before_fn()` and `after_fn()` for executing before / after each test.